server/backends/chunkstore/chunkstore.go

```diff
@@ -388,164 +394,185 @@ type ChunkstoreWriter struct {
+					// automatic flushes).
+					volatileTail = req.VolatileTail
+				}
+				chunk = append(chunk, req.Chunk...)
+				lastWriteSize = len(req.Chunk)
+				if req.Close {
+					open = false
 				}
-				l.chunk = append(l.chunk, req.Chunk...)
-				l.lastWriteSize = len(req.Chunk)
 			}
+		case <-time.After(time.Until(flushTime)):
+			// We timed out waiting for a write request. Flush the contents of the
+			// chunk to disk.
+			timeout = true
+		case <-ctx.Done():
+			open = false
+
+			// The context was canceled before the file was closed; extend the life of
+			// the context so we can flush the contents of the chunk and tail to disk.
+			var cancel func()
+			ctx, cancel = background.ExtendContextForFinalization(ctx, time.Second*10)
+			defer cancel()
 		}
-		return req
-	case <-time.After(time.Until(l.flushTime)):
-		l.timeout = true
-		return nil
-	case <-l.ctx.Done():
-		l.open = false
-		return nil
-	}
-}
 
-func (l *writeLoop) run() {
-	l.flushTime = time.Now().Add(year)
-	l.open = true
-	for l.open {
-		// Get the read request and update loop for this iteration.
-		req := l.readFromWriteChannel()
-		if !l.open {
+		if !open {
 			// We are closing the file, so any tail must be written lest it be lost.
-			l.chunk = append(l.chunk, l.volatileTail...)
-			l.lastWriteSize = len(l.volatileTail)
+			chunk = append(chunk, volatileTail...)
+			lastWriteSize = len(volatileTail)
+			volatileTail = []byte{}
 		}
 		var err error
-		for err == nil && len(l.chunk) >= l.writeBlockSize {
+		for err == nil && len(chunk) >= l.writeBlockSize {
 			// Write blocks until we have fewer than writeBlockSize bytes remaining
 			// or we encounter a write error.
-			err = l.write()
+			var n int
+			n, err = l.write(ctx, chunk, chunkIndex, lastWriteSize)
+			if err == nil {
+				bytesFlushed += n
+				chunk = chunk[n:]
+				chunkIndex++
+			}
 		}
 		// We do not attempt to write if we have already encountered a write error
 		// in this iteration of the loop, and instead we simply move on.
 		// If the file is empty and we are closing it, we must write an empty chunk.
 		// Otherwise, if we have data cached to write and we are either flushing or
 		// closing the file, we need to write a chunk.
-		if err == nil && ((!l.open && l.chunkIndex == 0) || (len(l.chunk) > 0 && (time.Now().After(l.flushTime) || !l.open))) {
-			l.write()
+		if err == nil && ((!open && chunkIndex == 0) || (len(chunk) > 0 && (time.Now().After(flushTime) || !open))) {
+			var n int
+			n, err = l.write(ctx, chunk, chunkIndex, lastWriteSize)
+			if err == nil {
+				bytesFlushed += n
+				chunk = chunk[n:]
+				chunkIndex++
+			}
 		}
-		if len(l.chunk) == 0 {
+		if len(chunk) == 0 {
 			// There is no data to write; we don't need to wake to flush data.
-			l.flushTime = time.Now().Add(year)
-		} else if time.Until(l.flushTime) > (l.writeTimeoutDuration) {
+			flushTime = time.Now().Add(year)
+		} else if time.Now().Add(l.writeTimeoutDuration).Before(flushTime) || bytesFlushed > 0 {
```

Comment:
This is the bug fix; basically we were only resetting the flushTime when we flushed the whole buffer to disk, not every time we wrote. This reflows the time part of the conditional to be identical, but hopefully more readable, and adds the condition that we should reset the flush time when we write to disk.

Reply:
@bduffany